garrysmod/lua/includes/extensions/vector.lua

```diff
@@ -3,10 +3,262 @@ local meta = FindMetaTable( "Vector" )
+function meta:GetNegated(...)
+	local v = Vector(self)
+	v:Negate(...)
+	return v
+end
+
+--[[---------------------------------------------------------
+Convets the vector to nagated version of itself
+-----------------------------------------------------------]]
+function meta:Nudge(num, vec)
+	self:Add(vec:GetMul(num))
+end
+
+--[[---------------------------------------------------------
+Returns a copy of the nagated vector
+-----------------------------------------------------------]]
+function meta:GetNudge(...)
+	local v = Vector(self)
+	v:Nudge(...)
+	return v
+end
 
+--[[---------------------------------------------------------
+Convets the vector to resized version of itself
+-----------------------------------------------------------]]
+function meta:Resize(num)
+	self:Normalize()
+	self:Mul(num)
+end
+
+--[[---------------------------------------------------------
+Returns a copy of the resized vector
+-----------------------------------------------------------]]
+function meta:GetResized(...)
+	local v = Vector(self)
+	v:Resize(...)
+	return v
+end
+
+--[[---------------------------------------------------------
+Convets the vector to resized midpoint of itself
+-----------------------------------------------------------]]
+function meta:Midpoint(vec)
+	self:Add(vec)
+	self:Mul(0.5)
+end
+
+--[[---------------------------------------------------------
+Returns a copy of the midpoint vector
+-----------------------------------------------------------]]
+function meta:GetMidpoint(...)
+	local v = Vector(self)
+	v:Midpoint(...)
+	return v
+end
+
+--[[---------------------------------------------------------
+Returns a copy of the zero point vector
+-----------------------------------------------------------]]
+function meta:GetZero(...)
+	local v = Vector(self)
+	v:Zero(...)
+	return v
+end
+
+--[[---------------------------------------------------------
+Applies new basis to a vector by the three axises of an angle
+-----------------------------------------------------------]]
+function meta:Basis(ang)
+	local x = self:Dot(ang:Forward())
+	local y = self:Dot(ang:Right())
+	local z = self:Dot(ang:Up())
+	self:SetUnpacked(x, y, z)
+end
+
+--[[---------------------------------------------------------
+Applies new basis to a vector by the three axises of an angle
+-----------------------------------------------------------]]
+function meta:GetBasis(...)
+	local v = Vector(self)
+	v:Basis(...)
+	return v
+end
+
+--[[---------------------------------------------------------
+Projects the vector in to an axis defined by the second vector
+-----------------------------------------------------------]]
+function meta:Project(vec)
+	local vn = vec:GetNormalized()
+	local nm = self:Dot(vn)
+	self:Set(vn)
+	self:Mul(nm)
+end
+
+--[[---------------------------------------------------------
+Projects the vector in to an axis defined by the second vector
+-----------------------------------------------------------]]
+function meta:GetProject(...)

```

Comment:
This must modify `self` this is different than `GetProject` which makes a copy. ;)

Reply:
This is done for better support where `A:getSomething` utilizes `A:Someting` and returns a copy ;)

Reply:
Written that way you only need to change the code in `meta:Project(...)` to adjust the behavior.

Reply:
It may be less duplicitous to have ``Get*`` call down to the non-returning alternatives, but it's slower. I guess it's not that big of a deal anyway.

Reply:
It may be less duplicitous to have Get* call down to the non-returning alternatives, but it's slower. I guess it's not that big of a deal anyway.

Reply:
Yes. The performance difference is negligible and you only need to change one method in the future instead of both. If `Project` works, then `GetProject` works too ;)